certora/specs/ConsistentState.spec

```
@@ -155,5 +170,92 @@ function hasSupplyCapIsEnabled(MetaMorphoHarness.Id id) returns bool {
     return supplyCap > 0 => enabled;
 }
 
+// Check that having a positive supply cap implies that the market is enabled.
+// This invariant is useful to conclude that market that are not enabled cannot be interacted with (notably for reallocate).
 invariant supplyCapIsEnabled(MetaMorphoHarness.Id id)
     hasSupplyCapIsEnabled(id);
+
+function hasDistinctIdentifiers(uint256 i, uint256 j) returns bool {
+    return i != j => withdrawQueue(i) != withdrawQueue(j);
+}
+
+// Check that there are no duplicate markets in the withdraw queue.
+invariant distinctIdentifiers(uint256 i, uint256 j)
+    hasDistinctIdentifiers(i, j)
+{
+    preserved updateWithdrawQueue(uint256[] indexes) with (env e) {
+        require hasDistinctIdentifiers(indexes[i], indexes[j]);
+    }
+}
+
+function isInWithdrawQueueIsEnabled(uint256 i) returns bool {
+    if(i >= withdrawQueueLength()) return true;
+
+    MetaMorphoHarness.Id id = withdrawQueue(i);
+    bool enabled;
+    _, enabled, _ = config(id);
+
+    return enabled;
+}
+
+// Check that markets in the withdraw queue are enabled.
+invariant inWithdrawQueueIsEnabled(uint256 i)
+    isInWithdrawQueueIsEnabled(i)
+filtered {
+    f -> f.selector != sig:updateWithdrawQueue(uint256[]).selector
+}
+
+rule inWithdrawQueueIsEnabledPreservedUpdateWithdrawQueue(env e, uint256 i, uint256[] indexes) {
+    uint256 j;
+    require isInWithdrawQueueIsEnabled(indexes[i]);
+
+    requireInvariant distinctIdentifiers(indexes[i], j);
+
+    updateWithdrawQueue(e, indexes);
+
+    MetaMorphoHarness.Id id = withdrawQueue(i);
```

could we take that before (I find it clearer)
```suggestion
    MetaMorphoHarness.Id id = withdrawQueue(indexes[i]);
```

---

Yes we could, but the meaning of `id` is the identifier that we check the property on at the end (with the `assert`), so I don't think it will make that clearer

---

Keeping it like that makes it clearer that `id`is the same as the one defined in the function `isInWithdrawQueueIsEnabled`